src/Icicle/Avalanche/Statement/Simp.hs

```diff
@@ -391,10 +430,16 @@ substXinS a_fresh name payload statements
 --      statements that do not have any external effect are silly
 --  * Constant folding for ifs
 --
-thresherNoAlpha :: (Hashable n, Eq n, Ord a) => a -> Statement a n Prim -> FixT (Fresh n) (Statement a n Prim)
+thresherNoAlpha
+ :: (Hashable n, Eq n, Ord a)
+ => a
+ -> Statement (Ann a n) n Prim
+ -> FixT (Fresh n) (Statement (Ann a n) n Prim)
 thresherNoAlpha a_fresh statements
  = transformUDStmt trans Map.empty statements
  where
+  xVar n
+   = XVar (a_fresh, Set.singleton n) n
   trans env s
    -- Check if it actually does anything:
    -- updates accumulators, returns a value, etc.
```

Comment:
I think this "hasEffect" might end up being the bottleneck soon. it might be faster in a completely separate transform to thresher, since thresher is top-down and hasEffect seems to be bottom-up.
you could perhaps put it in dead (it does make sense), but dead stuffs around with the fixpoint of loops, so hasEffect might be better as a separate pass since it doesn't need loop fixpoints.